src/web_analysis/test_robots.py

```diff
@@ -0,0 +1,1087 @@
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "JPG images in the '/images' directory should be blocked.",
+        )
+
+    def test_allow_overrides_disallow(self):
+        """Tests that the Allow directive properly overrides a conflicting Disallow for the same path."""
+        agent_rules = {"*": {"Disallow": ["/folder"], "Allow": ["/folder/subfolder"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Allow directive should override conflicting Disallow for the same path, ensuring access to /folder/subfolder.",
+        )
+
+    def test_empty_user_agent_and_path(self):
+        """Tests behavior when user-agent and path are explicitly empty, which should imply no restrictions."""
+        agent_rules = {"": {"Disallow": [""]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "none", "GoogleBot": "none"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "An empty user-agent and path in Disallow should lead to no restrictions.",
+        )
+
+    def test_unicode_path_handling(self):
+        """Tests the function's ability to handle paths with Unicode characters."""
+        agent_rules = {"*": {"Disallow": ["/über"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Paths with Unicode characters should be properly disallowed.",
+        )
+
+    def test_multiple_disallow_with_wildcards(self):
+        """Tests handling of multiple Disallow entries with wildcards for a single agent."""
+        agent_rules = {"GoogleBot": {"Disallow": ["/tmp/*", "/private/*/config"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "none", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Multiple Disallow entries with wildcards should be correctly handled for GoogleBot.",
+        )
+
+    def test_conflicting_allow_disallow_same_path(self):
+        """Tests resolution of conflicting Allow and Disallow rules for the exact same path."""
+        agent_rules = {"*": {"Disallow": ["/data"], "Allow": ["/data"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "none", "GoogleBot": "none"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Conflicting Allow and Disallow for the same path should default to allowing access.",
+        )
+
+    def test_specific_vs_generic_user_agents(self):
```

Comment:
This should also be GoogleBot "none" I believe?

https://developers.google.com/search/docs/crawling-indexing/robots/robots_txt#order-of-precedence-for-user-agents

Reply:
This is a confusing one. So as far as I think if GoogleBot has a specific allow rule that contradicts a broader disallow rule for * then the specific rule for GoogleBot should take precedence in that scenario?

"*" has a Disallow directive for /private.
GoogleBot has an Allow directive for /private/stats so it takes precedence?